### PR TITLE
rpc: avoid needless locking

### PR DESCRIPTION
**Pull request by tamird**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8649)

### CODE REVIEW

File: rpc/context.go [hunk 1]
```diff
@@ -146,17 +146,18 @@ func (ctx *Context) SetLocalInternalServer(internalServer roachpb.InternalServer
...
+	ctx.removeConnLocked(key, meta)
 }
 
-func (ctx *Context) removeConnLocked(key string, conn *grpc.ClientConn) {
+func (ctx *Context) removeConnLocked(key string, meta *connMeta) {
 	if log.V(1) {
 		log.Infof(ctx.masterCtx, "closing %s", key)
 	}
-	if conn != nil {
+	meta.Do(func() {}) // Make sure initialization is not in-progress.
```
Comment (spencerkimball) [inline rpc/context.go]: This seems problematic as it'll be reentrant on the `Sync.Once` mutex if called during error handling from the real `Do()` invocation. You should probably move this to the shutdown worker.

File: rpc/context.go [hunk 2]
```diff
@@ -191,27 +192,23 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
 			log.Infof(ctx.masterCtx, "dialing %s", target)
 		}
 		conn, err := grpc.DialContext(ctx.masterCtx, target, dialOpts...)
```
Comment (spencerkimball) [inline rpc/context.go]: change this back to:

`meta.conn, meta.err = grpc.DialContext(ctx.masterCtx, target, dialOpts...)`

File: rpc/context.go [hunk 3]
```diff
@@ -191,27 +192,23 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
...
-		ctx.conns.Lock()
 		meta.conn = conn
```
Comment (spencerkimball) [inline rpc/context.go]: And remove these assignments.

**Comment by spencerkimball** [comment]
LGTM; make sure you run stressrace on this.

File: rpc/context.go (hunk 1 above)
Comment (tamird) [inline rpc/context.go]: That error handling code runs in a different goroutine.

Comment (tamird) [inline rpc/context.go]: Oh, I see what you're saying. Looking.

File: rpc/context.go (hunk 3 above)
Comment (tamird) [inline rpc/context.go]: Done.

File: rpc/context.go (hunk 2 above)
Comment (tamird) [inline rpc/context.go]: Done.

File: rpc/context.go (hunk 1 above)
Comment (tamird) [inline rpc/context.go]: Done.

**Comment by tamird** [comment]
Will do. Any tests in particular?

**Comment by spencerkimball** [comment]
Unfortunately I can't remember which test blew up on race last time.

**Pull request merged**